### ISSUE TITLE
scalapack interface seems broken?

### ISSUE THREAD

**Issue by QiaoLei-88** [dealii/dealii#7534]
When compiling 17abd6c7e1 on CentOS 7 with gcc7, I run into the following error: 

```
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc: In member function ‘void dealii::ScaLAPACKMatrix<NumberType>::save_parallel(const string&, const std::pair<unsigned int, unsigned int>&) const’:
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2854:20: error: there are no arguments to ‘H5Pset_fapl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_fapl_mpio’ must be available [-fpermissive]
   status         = H5Pset_fapl_mpio(plist_id, tmp.grid->mpi_communicator, info);
                    ^~~~~~~~~~~~~~~~
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2854:20: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2931:14: error: there are no arguments to ‘H5Pset_dxpl_mpio’ that depend on a template parameter, so a declaration of ‘H5Pset_dxpl_mpio’ must be available [-fpermissive]
   status   = H5Pset_dxpl_mpio(plist_id, H5FD_MPIO_INDEPENDENT);
              ^~~~~~~~~~~~~~~~
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc: In instantiation of ‘void dealii::ScaLAPACKMatrix<NumberType>::save_parallel(const string&, const std::pair<unsigned int, unsigned int>&) const [with NumberType = double; std::string = std::basic_string<char>]’:
/home/qiaol/package_install/deal.II-git/build_devel_gcc7/source/lac/scalapack.inst:10:16:   required from here
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2854:36: error: ‘H5Pset_fapl_mpio’ was not declared in this scope
   status         = H5Pset_fapl_mpio(plist_id, tmp.grid->mpi_communicator, info);
                    ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2854:36: note: suggested alternative: ‘H5Pset_fapl_stdio’
   status         = H5Pset_fapl_mpio(plist_id, tmp.grid->mpi_communicator, info);
                    ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                    H5Pset_fapl_stdio
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2931:30: error: ‘H5Pset_dxpl_mpio’ was not declared in this scope
   status   = H5Pset_dxpl_mpio(plist_id, H5FD_MPIO_INDEPENDENT);
              ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/qiaol/devel/dealii_git/source/lac/scalapack.cc:2931:30: note: suggested alternative: ‘H5Pset_fapl_stdio’
   status   = H5Pset_dxpl_mpio(plist_id, H5FD_MPIO_INDEPENDENT);
              ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
              H5Pset_fapl_stdio
```

The same code is compiled successfully on my Mac with clang 10. So I am not sure whether this is a deal.II problem or some thing wrong in my pre-built third party library.

Is the prefix "H5" suggests something related to HDF5 library?

**Comment by davydden** [comment]
> Is the prefix "H5" suggests something related to HDF5 library?

yes, those are functions to save Scalapack using HDF5. It looks like you build with `HDF5` but the version you have does not have the functions used here. 

> The same code is compiled successfully on my Mac with clang 10. So I am not sure whether this is a deal.II problem or some thing wrong in my pre-built third party library.

which HDF5 versions you have on Centos and on macOS?

**Comment by masterleinad** [comment]
Do you know which `HDF5` version you are using? According to the changelog http://davis.lbl.gov/Manuals/HDF5-1.6.1/ADGuide/Changes.html, `H5Pset_fapl_mpio` was introduced in version 1.4.0 released before 2003-07-03 according to https://portal.hdfgroup.org/display/support/Downloads...

**Comment by masterleinad** [comment]
For me, `H5FDmpio.h` is included by `H5FDmpi.h` which is included by `hdf5.h`.
In `deal.II`, `hdf5.h` is included in both `source/base/data_out_base.cc`
`source/lac/scalapack.cc` guarded by a preprocessor switch checking if `HDF5` support is enabled.

Do you see the same? 
How do the two configurations differ otherwise? Can you provide the content of `detailed.log` for both?

**Comment by masterleinad** [comment]
Unfortunately, that doesn't reveal as much as I have hoped.
Can you check that the inclusion chain I described [above](https://github.com/dealii/dealii/issues/7534#issuecomment-447658711) is the same for you?

**Comment by QiaoLei-88** [comment]
> Can you check that the inclusion chain I described above is the same for you?

I have checked the inclusion chain and it is the same with your description. 
I found that `#include<hdf5.h>` is active only when `DEAL_II_WITH_HDF5` is defined and `H5FDmpio.h` needs `H5_HAVE_PARALLEL`.
Although these two macros should be defined as configured, is there any way to check preprocessed code in deal.II conveniently?

**Comment by masterleinad** [comment]
`include/deal.II/base/config.h` in you build or install directory should contain the line:
```
#define DEAL_II_WITH_HDF5
```
Given that this is the first error you are encountering, this preprocessor variable is likely defined.
With respect to the `HDF5` files `include/H5pubconf.h` contains the line
```
#define H5_HAVE_PARALLEL 1
```
for me. In particular, we check for the existence of this preprocessor variable in `cmake/modules/FindHDF5.cmake` and in `cmake/configure/configure_hdf5.cmake` we check that either both `deal.II` and `HDF5` are configured with `MPI` support or none of them is.

Can you check you `include/H5pubconf.h`?

**Comment by QiaoLei-88** [comment]
> Can you check you `include/H5pubconf.h`?

`#define H5_HAVE_PARALLEL 1` exists in `include/H5pubconf.h`.

This makes nonsense. It must another header file be included.

Can I pass `-E` option to make dealii compiling stop right after preprocess?

**Comment by masterleinad** [comment]
I would just invoke `make` in verbose mode, i.e. call `make VERBOSE=1` and extract the failing compilation unit. You should be able to see the exact line that is used for compiling. Then you can try to remove unrelated include directories to narrow down the problematic ones.
Of course, you can also check in `source/lac/scalapack.cc` after including all headers whether `H5_HAVE_PARALLEL` is defined or not by using a `static_assert` (or trying to define `H5_HAVE_PARALLEL` in that file on your own).

**Comment by QiaoLei-88** [comment]
After check the preprocessed `source/lac/scalapack.cc`, I found out the `hdf5.h` is included from `Anaconda3-5.2.0` installation, which has no mpi support.

The include path to `Anaconda3-5.2.0` is introduced by searching for ZLIB
```
-- Include /home/qiaol/devel/dealii_git/cmake/configure/configure_1_zlib.cmake
-- Found ZLIB: /opt/Anaconda3-5.2.0/lib/libz.so (found version "1.2.11") 
--   ZLIB_VERSION: 1.2.11
--   ZLIB_LIBRARIES: /opt/Anaconda3-5.2.0/lib/libz.so
--   ZLIB_INCLUDE_DIRS: /opt/Anaconda3-5.2.0/include
```

Dealii is successfully built after disabled ZLIB by `-DDEAL_II_WITH_ZLIB=OFF`.

But I can't find any clue that how dealii knows the existence of `Anaconda3-5.2.0`.
There is no `Anaconda3-5.2.0` in `$INCLUDE`, `$CPATH` nor CMAKE options for dealii building.

**Comment by davydden** [comment]
> The include path to Anaconda3-5.2.0 is introduced by searching for ZLIB

In this case I am tempted to close the issue as there is not much we can do here. 

> But I can't find any clue that how dealii knows the existence of Anaconda3-5.2.0

Interesting indeed, maybe CMake automatically checks those paths when searching for packages.
